### PR TITLE
Code Review request for Tonin and Scialpi

### PR DESCRIPTION
**Pull request by ricdezen**
Good day, I would need a code review for the following files:
- PingPendingRequest
- FindNodePendingRequest
- PendingRequestManager

The other attached files are interfaces regarding the functionality of said files. There is little to review there but you are free to.
PendingRequests are the way in which I decided to keep track of all the various operations the Kademlia network undergoes. I included only the Ping and FindNode requests because they are the best examples in my opinion. An Invite is almost identical to a Ping and both Find Value and Store add very little to the basic Node lookup.
I'd like to apologize in advance, because due to various reasons, group 1 hasn't reached a full implementation of Kademlia yet and some parts of our architecture still need to be refined or are stubbed entirely. This means the code doesn't have proper integration testing, and my classes in particular stand in an awful spot, where Unit testing itself proves very difficult since it depends on every other aspect of the Network. As such, the actual correctness of FindNodePendingRequest is still to be proven.

### CODE REVIEW

**Comment by RicDeZen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/FindNodePendingRequest.java  4
- kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PingPendingRequest.java  2
- kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PendingRequestManager.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/RicDeZen/ingsw-group1-kademlia/pullRequest?prid=4743795&bid=15830564)

File: kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PendingRequestManager.java [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+     * considered.
+     *
+     * @param action The action continuing the corresponding Request
+     * @return true if a matching Request was found and executed, false otherwise.
+     * @throws InvalidActionException If {@code action} is not a Response type Action.
+     */
+    public boolean continueRequest(KadAction action) {
+        if (!action.getActionType().isResponse())
+            throw new InvalidActionException(CONTINUE_ERR);
+        int operationId = action.getOperationId();
```
Comment (RicDeZen) [inline kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PendingRequestManager.java]: Codacy Issue found: [Avoid unused local variables such as 'operationId'.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-kademlia/pullRequest?prid=4743795)

File: kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/FindNodePendingRequest.java [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+
+    //These constants are soon to be moved to an appropriate class.
+    /**
+     * The number of Nodes contacted on each iteration and the number of Nodes every contacted
+     * Node will send back.
+     */
+    private static final int KAD_K = 10;
+    /**
+     * The length of IDs in the Network and the number of buckets in the Routing Table.
+     */
```
Comment (tonin-alessandra) [inline kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/FindNodePendingRequest.java:43]: Maybe you can specify the unit of measurement of the length of the ID's: for me it is clear because I'm working on this project, but for other people it could not be .

File: kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/FindNodePendingRequest.java [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+
+    /**
+     * Default constructor.
+     *
+     * @param operationId      the unique id for this PendingRequest operation.
+     * @param targetId         the id of the Node we are looking for.
+     * @param actionPropagator a valid {@link ActionPropagator}.
+     * @param nodeProvider     a valid {@link NodeDataProvider}.
+     * @param resultListener   a valid listener to this {@code PendingRequest}'s Result.
+     */
```
Comment (tonin-alessandra) [inline kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/FindNodePendingRequest.java:78]: This method has 5 params, it is a quite big number..is there any way to reduce them?
Maybe no, otherwise you would have already done it, I think.

File: kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/FindNodePendingRequest.java [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+     * @see PendingRequest#getTotalStepsTaken()
+     */
+    @Override
+    public int getTotalStepsTaken() {
+        return totalStepsTaken;
+    }
+
+    /**
+     * @see PendingRequest#getOperationId()
+     */
```
Comment (tonin-alessandra) [inline kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/FindNodePendingRequest.java:105]: You forgot the @return specification.

File: kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/FindNodePendingRequest.java [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+     * @return true if the given action can be used to continue the operation, false otherwise.
+     * The action is always ignored if the current state is not
+     * {@link RequestState#PENDING_RESPONSES}.
+     * The action is "pertinent" if:
+     * - The {@code ActionType} of {@code action} is
+     * {@link ingsw.group1.kademlia.KadAction.ActionType#FIND_NODE_ANSWER}.
+     * - The {@code operationId} matches.
+     * - The Payload is either a
+     * @see PendingRequest#isActionPertinent(KadAction)
+     */
```
Comment (tonin-alessandra) [inline kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/FindNodePendingRequest.java:142]: I think you forgot the @param description.

File: kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PendingRequestManager.java [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+     * Constructor.
+     *
+     * @param capacity                The maximum amount of PendingRequests that can be pending
+     *                                at a time. Can't
+     *                                be higher than {@link PendingRequestManager#MAX_SIZE}.
+     * @param defaultActionPropagator The {@link ActionPropagator} that will be used by default
+     *                                by enqueued {@code PendingRequests}.
+     * @param defaultNodeDataProvider The {@link NodeDataProvider} that will be used by default
+     *                                by enqueued {@code PendingRequests}.
+     */
```
Comment (tonin-alessandra) [inline kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PendingRequestManager.java:54]: Missing @throws specification.

File: kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PendingRequestManager.java [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+    public void setDefaultNodeDataProvider(@NonNull NodeDataProvider<BinarySet, PeerNode> newDefaultProvider) {
+        defaultNodeDataProvider = newDefaultProvider;
+    }
+
+    /**
+     * Method to enqueue a {@link PingPendingRequest}.
+     *
+     * @return true if the {@code PendingRequest} could be enqueued, false otherwise.
+     */
```
Comment (tonin-alessandra) [inline kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PendingRequestManager.java:91]: Missing @param statements.
Probably variables' names are self-explicative, but I think it is always better to specify params description.

File: kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PendingRequestManager.java [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+        currentRequests.add(request);
+        request.start();
+        return true;
+    }
+
+    /**
+     * Method to enqueue a {@link InvitePendingRequest}.
+     *
+     * @return true if the {@code PendingRequest} could be enqueued, false otherwise.
+     */
```
Comment (tonin-alessandra) [inline kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PendingRequestManager.java:114]: Same as enqueuePing.

File: kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PendingRequestManager.java [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+        currentRequests.add(request);
+        request.start();
+        return true;
+    }
+
+    /**
+     * Method to enqueue a {@link FindNodePendingRequest}.
+     *
+     * @return true if the {@code PendingRequest} could be enqueued, false otherwise.
+     */
```
Comment (tonin-alessandra) [inline kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PendingRequestManager.java:138]: Same as enqueuePing.

File: kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PendingRequestManager.java [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+        currentRequests.add(request);
+        request.start();
+        return true;
+    }
+
+    /**
+     * Method to enqueue a {@link FindValuePendingRequest}.
+     *
+     * @return true if the {@code PendingRequest} could be enqueued, false otherwise.
+     */
```
Comment (tonin-alessandra) [inline kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PendingRequestManager.java:162]: Same as enqueuePing.

File: kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PendingRequestManager.java [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+        currentRequests.add(request);
+        request.start();
+        return true;
+    }
+
+    /**
+     * Method to enqueue a {@link StorePendingRequest}.
+     *
+     * @return true if the {@code PendingRequest} could be enqueued, false otherwise.
+     */
```
Comment (tonin-alessandra) [inline kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PendingRequestManager.java:186]: Same as enqueuePing.

File: kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PingPendingRequest.java [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+
+    /**
+     * Default constructor.
+     *
+     * @param operationId      the unique id for this PendingRequest operation.
+     * @param peerToPing       the {@code Peer} to ping.
+     * @param actionPropagator a valid {@link ActionPropagator}.
+     * @param nodeProvider     a valid {@link NodeDataProvider}.
+     * @param resultListener   a valid listener to this {@code PendingRequest}'s Result.
+     */
```
Comment (tonin-alessandra) [inline kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PingPendingRequest.java:45]: Same as FindNodePendingRequest, there are 5 params here.

File: kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PingPendingRequest.java [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+     * @see PendingRequest#getTotalStepsTaken()
+     */
+    @Override
+    public int getTotalStepsTaken() {
+        return totalStepsTaken;
+    }
+
+    /**
+     * @see PendingRequest#getOperationId()
+     */
```
Comment (tonin-alessandra) [inline kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PingPendingRequest.java:74]: Missing @return.

File: kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PingPendingRequest.java [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+     * The action is always ignored if the current state is not
+     * {@link RequestState#PENDING_RESPONSES}.
+     * The action is "pertinent" if:
+     * - The {@code ActionType} of {@code action} is
+     * {@link KadAction.ActionType#PING_ANSWER}.
+     * - The {@code operationId} matches.
+     * - The {@code Peer} of {@code action} matches the {@code Peer} of
+     * {@link PingPendingRequest#pingAction}.
+     * @see PendingRequest#isActionPertinent(KadAction)
+     */
```
Comment (tonin-alessandra) [inline kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PingPendingRequest.java:110]: Missing @param specification.

**Review by tonin-alessandra — CHANGES_REQUESTED** [review]
What to say, your code is really really good, well done!
Your code is advanced and not so easy to completely understand for me, but you did a really good job. In particular, I liked a lot the fact that you always explained everything that could be not totally clear, this made your code easier to read (for example, variables' names, classes' descriptions...).
I made some comments about specifications, sometimes you forgot the @return or the @param specs, but this is not a big problem.
In general:

- your code is well-formatted
- methods/variables names are clear and consistent
- there's no hardwiring
- no TODO/fixme
- the number of methods' params is correct, except for two specific methods I commented (maybe it isn't an error, must be checked)
- code documentation is complete and clear, except on specific points where specs could be written better

Now I'd like to give you a couple of suggestions if I can:
1) in my opinion, it is nicer to write ID than id;
2) I saw that all "enqueueX" methods contain similar code: maybe is it possible to extract it in a common method, without rewriting the same code every time? Or maybe you should write a unique method for each type of enqueue, and inside it you can distinguish different cases (?).
I'm not totally sure about the correctness of this second statement, I know you're more clever than me in coding so I think your solution is better than mine (and probably mine is not achievable, too), but I had this doubt so I wanted to tell you.
In conclusion, I learned a lot reviewing your code, I have nothing to say except good work, keep going like this!

**Comment by ricdezen** [comment]
You are right, I missed lots of @param, I'll fix them tomorrow or later today, thanks!

File: kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/FindNodePendingRequest.java (hunk 2 above)
Comment (ricdezen) [reply kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/FindNodePendingRequest.java:43]: Good point, I should change this

File: kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/FindNodePendingRequest.java (hunk 3 above)
Comment (ricdezen) [reply kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/FindNodePendingRequest.java:78]: I agree 5 parameters are way too many, there used to be a Factory class for PendingRequests, which reduced the number of required to two and allowed PendingRequest subclasses to be package private, but I had to get rid of it because of trouble with initializing some types of Requests. The removal of the factory is also the same reason for the PendingRequestManager having 5 methods with almost identical bodies, he kinda took the place of the Factory. I'll definitely be reducing the parameters in the near future, when we will implement some Object Pools for most classes in the network, so every Pending Request can just receive some sort of key and ask for the right instance by itself.

**Comment by ricdezen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/FindNodePendingRequest.java  4
- kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PendingRequestManager.java  5
- kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PingPendingRequest.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/RicDeZen/ingsw-group1-kademlia/pullRequest?prid=4743795&bid=15830564)

File: kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PendingRequestManager.java [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+     * considered.
+     *
+     * @param action The action continuing the corresponding Request
+     * @return true if a matching Request was found and executed, false otherwise.
+     * @throws InvalidActionException If {@code action} is not a Response type Action.
+     */
+    public boolean continueRequest(KadAction action) {
+        if (!action.getActionType().isResponse())
+            throw new InvalidActionException(CONTINUE_ERR);
+        int operationId = action.getOperationId();
```
Comment (ricdezen) [inline kademlia/src/main/java/ingsw/group1/kademlia/pendingrequests/PendingRequestManager.java:215]: Codacy Issue found: [Avoid unused local variables such as 'operationId'.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-kademlia/pullRequest?prid=4743795)
